src/core/common/settings.cpp

```diff
@@ -0,0 +1,508 @@
+}
+
+ThreadError otPlatSettingsSet(uint16_t aKey, const uint8_t *aValue, int aValueLength)
+{
+    ThreadError error = kThreadError_None;
+    uint32_t address = sBase + kSettingsFlagSize;
+
+    while (address < (sBase + sSettingsSize))
+    {
+        struct settingsBlock block;
+
+        otPlatFlashRead(address, reinterpret_cast<uint8_t *>(&block), sizeof(block));
+
+        if ((!(block.flag & kBlockAddFlag) && block.flag & kBlockDeleteFlag) && (block.key == aKey))
+        {
+            SuccessOrExit(error = otPlatSettingsDelete(aKey, -1));
+            break;
+        }
+
+        address += (block.length + sizeof(struct settingsBlock));
+    }
+
+    error = otPlatSettingsAdd(aKey, aValue, aValueLength);
+
+exit:
+    return error;
+}
+
+ThreadError otPlatSettingsAdd(uint16_t aKey, const uint8_t *aValue, int aValueLength)
+{
+    ThreadError error = kThreadError_None;
+    struct stageAddSettingsBlock *stageBlock;
+    int index = -1;
+    uint32_t address = sBase + kSettingsFlagSize;
+
+    VerifyOrExit((!sCommitLock) || (sCommitLock &&
+                                    (sStageAddSettingsNum < kMaxStageAddNum)), error = kThreadError_NoBufs);
+
+    stageBlock = &sStageAddSettingsBlock[sStageAddSettingsNum];
+    stageBlock->block.key = aKey;
+
+    while (address < (sBase + sSettingsSize))
+    {
+        struct settingsBlock block;
+
+        otPlatFlashRead(address, reinterpret_cast<uint8_t *>(&block), sizeof(block));
+
+        if (!(block.flag & kBlockAddFlag) && (block.flag & kBlockDeleteFlag) && (block.key == aKey) && (block.index > index))
+        {
+            index = block.index;
+        }
+        else if (block.flag == 0xff)
+        {
+            break;
+        }
+
+        address += (block.length + sizeof(struct settingsBlock));
+    }
+
+    stageBlock->block.index = static_cast<uint8_t>(index + 1);
+
+    if (sCommitLock)
+    {
+        for (index = 0; index < sStageAddSettingsNum; index++)
+        {
+            if ((sStageAddSettingsBlock[index].block.key == aKey) &&
+                (sStageAddSettingsBlock[index].block.index >= stageBlock->block.index))
+            {
+                stageBlock->block.index = sStageAddSettingsBlock[index].block.index + 1;
+            }
+        }
+    }
+
+    stageBlock->block.flag &= (~kBlockAddFlag);
+    stageBlock->block.length = static_cast<uint16_t>(aValueLength);
+    memcpy(stageBlock->data, aValue, stageBlock->block.length);
+
+    // padding
+    while (stageBlock->block.length & 3)
+    {
+        stageBlock->data[stageBlock->block.length++] = 0xff;
+    }
+
+    if (!sCommitLock)
+    {
+        if ((sUsedSize + stageBlock->block.length + sizeof(struct settingsBlock)) >= sSettingsSize)
+        {
+            VerifyOrExit(swapSettingsBlock() >= (stageBlock->block.length + sizeof(struct settingsBlock)),
+                         error = kThreadError_NoBufs);
+        }
+
+        otPlatFlashWrite(sBase + sUsedSize, reinterpret_cast<uint8_t *>(stageBlock),
```

Comment:
I split kSettingsAdd flag into kBlockAddBeginFlag and kBlockAddCompleteFlag for atomic write.